### PR TITLE
feat(doc): documentation describing Kubewarden components

### PR DESCRIPTION
**Pull request by jvanz**
## Description

Adds a documentation file given some description about the Kubewarden components and how they work.

Fix https://github.com/kubewarden/kubewarden-controller/issues/686

### CODE REVIEW

**Review by robertsirc — APPROVED** [review]
This looks fine to me.

File: components.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
+# Kubewarden Components
+
+Kubewarden offers a secure and flexible approach to enforce policies within
+Kubernetes clusters, leveraging the speed and isolation of WebAssembly. This
+documentation delves into Kubewarden's architecture, outlining the roles,
+functionalities, and interconnectivity of its components, essential for
+deploying and utilizing Kubewarden effectively.
```
Comment (jhkrug) [inline components.md:7]:
```suggestion
Kubewarden offers a secure and flexible approach to enforcing policies within
Kubernetes clusters, using the speed and isolation of WebAssembly. This
documentation explores Kubewarden's architecture, outlining the roles,
functionalities, and interconnectivity of its components, essential for
deploying and utilizing Kubewarden effectively.
```

File: components.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+### Components of Kubewarden
+
+Kubewarden's architecture is comprised of meticulously designed components,
+each serving a distinct function within the ecosystem, ensuring that clusters
+adhere to defined operational policies through efficient and secure means.
```
Comment (jhkrug) [inline components.md:13]: Is this necessary? Is it for marketing purposes? It's very 'blurby'.

File: components.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+#### Kubewarden CRDs (Custom Resource Definitions)
+
+Kubewarden CRDs are pivotal in defining resources for admission request
+validations. They detail the structure for not only the
+`ClusterAdmissionPolicy` and `AdmissionPolicy` resources, which include policy
+configurations and matching rules for admission requests, but also for the
+`PolicyServer` resource. This addition delineates the setup for Policy Servers,
+encompassing their deployment specifics and operational parameters, thereby
+forming the backbone for policy management and execution within the ecosystem.
```
Comment (jhkrug) [inline components.md:23]:
```suggestion
Kubewarden CRDs define resources for admission request
validations. They detail the structure, not only for the
`ClusterAdmissionPolicy` and `AdmissionPolicy` resources, which include policy
configurations and matching rules for admission requests, but also for the
`PolicyServer` resource. CRDs define the setup for Policy Servers,
with deployment specific details and parameters.
```

File: components.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+repository.
+
+#### Kubewarden Controller
+
+The Kubewarden Controller stands as the orchestrator, closely monitoring
+Kubewarden CRDs and ensuring the Kubernetes resources necessary for running
+Policy Servers and policies are aptly configured. Beyond this, the controller
+extends its vigilance to some native Kubernetes resources essential for the
+seamless operation of Kubewarden resources, showcasing its integral role in
+maintaining the ecosystem's stability and responsiveness.
```
Comment (jhkrug) [inline components.md:36]:
```suggestion
The Kubewarden Controller is the conductor,  monitoring
Kubewarden CRDs and ensuring the Kubernetes resources necessary for running
Policy Servers and policies are correctly configured. Beyond this, the controller
monitors  Kubernetes resources 
used by Kubewarden resources, helping maintain system stability and performance.
```

File: components.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+repository.
+
+#### Policy Server
+
+Serving as the execution ground for policies, the Policy Server is tasked with
+the real-time processing of admission requests, utilizing the logic
+encapsulated within each policy to validate, mutate, or reject requests.
+Deployed within Kubernetes, these servers are engineered for high performance
+and isolation, capable of hosting multiple policies concurrently, emphasizing
+Kubewarden's commitment to security and efficiency.
```
Comment (jhkrug) [inline components.md:49]:
```suggestion
The Policy Server processes admission requests, using the logic
encapsulated within each policy to validate, mutate, or reject requests.
Deployed within Kubernetes, these servers are engineered for high performance
and isolation, capable of hosting multiple policies concurrently, emphasizing
Kubewarden's commitment to security and efficiency.
```

File: components.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+See the Policy Server code [here](https://github.com/kubewarden/policy-server)
+
+#### Policies
+
+At the heart of Kubewarden's operational logic lie the Policies—OCI artifacts
+that contain both the WebAssembly module and necessary metadata for execution.
+These policies embody the rules for admission control, packaged for
+distribution and executed within Policy Servers to ensure compliance with the
+defined security and operational standards.
```
Comment (jhkrug) [inline components.md:59]:
```suggestion
At the heart of Kubewarden's operational logic lies the Policies—OCI artifacts
that contain both the WebAssembly modules and necessary metadata for execution.
These policies embody the rules for admission control, packaged for
distribution and executed within Policy Servers to ensure compliance with the
defined security and operational standards.
```

File: components.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+[here](https://github.com/search?q=org%3Akubewarden%20topic%3Akubewarden-policy&type=repositories)
+
+#### ValidationWebhooks and MutatingWebhooks
+
+Kubewarden's integration with Kubernetes is orchestrated through
+ValidationWebhooks and MutatingWebhooks, mechanisms that funnel admission
+requests to Policy Servers. Crafted by the Kubewarden Controller, these
+webhooks are critical for directing traffic, ensuring that each admission
+request is evaluated against the relevant policy, facilitating a dynamic and
+automated policy enforcement process.
```
Comment (jhkrug) [inline components.md:71]:
```suggestion
Kubewarden's integration with Kubernetes is orchestrated through
ValidationWebhooks and MutatingWebhooks, mechanisms that funnel admission
requests to Policy Servers. Created by the Kubewarden Controller, these
webhooks are critical for directing traffic, ensuring that each admission
request is evaluated against the relevant policy, facilitating a dynamic and
automated policy enforcement process.
```

File: components.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+#### Audit Scanner
+
+The Audit Scanner introduces a proactive layer to Kubewarden’s security
+framework, periodically scanning the cluster to identify and report
+non-compliant resources. This component enhances cluster security by
+continuously evaluating the cluster state against enforced policies,
+pinpointing discrepancies and facilitating timely remediation. It integrates
+seamlessly with Policy Servers, utilizing existing policy definitions to ensure
+consistent enforcement and evaluation.
```
Comment (jhkrug) [inline components.md:81]: It's not possible to be continuous and periodic at the same time

```suggestion
The Audit Scanner introduces a proactive layer to Kubewarden’s security
framework, periodically scanning the cluster to identify and report
non-compliant resources. This component enhances cluster security by
 evaluating the cluster state against enforced policies,
identifying discrepancies and facilitating timely remediation. It integrates
seamlessly with Policy Servers, using existing policy definitions to ensure
consistent enforcement and evaluation.
```

**Review by jhkrug — COMMENTED** [review]
Some thoughts. Who's it for? Is it publicity material? Marketing material. Even if so, I'd tone it down a little, make it a little more technically focused language.

File: components.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
+# Kubewarden Components
+
+Kubewarden's architecture is comprised multiple components, each serving a
```
Comment (viccuad) [inline components.md:3]:
```suggestion
Kubewarden's architecture is comprised by multiple components, each serving a
```

File: components.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+distinct function in the Kubewarden stack. Ensuring users' compliance and
```
Comment (viccuad) [inline components.md:4]:
```suggestion
distinct function in the Kubewarden stack, ensuring users' compliance and
```

File: components.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+secuirty policies are properly enforced
```
Comment (viccuad) [inline components.md:5]:
```suggestion
secuirty policies are properly enforced.
```

File: components.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+The code of the Kubewarden controller can be found in the
+[kubewarden-controller](https://github.com/kubewarden/kubewarden-controller)
+repository.
+
+#### Policy Server
+
+The Policy Server processes admission requests, using the logic encapsulated
+within each policy to validate, mutate, or reject requests. Deployed within
+Kubernetes, these servers are capable of hosting multiple policies
+concurrently.
```
Comment (viccuad) [inline components.md:34]:
```suggestion
in parallel.
```

(given that we use threads).

File: components.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+See the Policy Server code [here](https://github.com/kubewarden/policy-server)
+
+#### Policies
+
+At the heart of Kubewarden's operational logic lies the Policies—OCI artifacts
```
Comment (viccuad) [inline components.md:40]:
```suggestion
At the heart of Kubewarden's operational logic lies the Policies — OCI artifacts
```

**Review by viccuad — APPROVED** [review]

File: components.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
+# Kubewarden Components
+
+Kubewarden's architecture is comprised by multiple components, each serving a
+distinct function in the Kubewarden stack, ensuring users' compliance and
+secuirty policies are properly enforced.
```
Comment (jhkrug) [inline components.md:5]:
```suggestion
Kubewarden's architecture comprises multiple components, each serving a
distinct function in the Kubewarden stack, ensuring users' compliance and
security policies are properly enforced.
```

File: components.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+The Policy Server processes admission requests, using the logic encapsulated
+within each policy to validate, mutate, or reject requests. Deployed within
+Kubernetes, these servers are capable of hosting multiple policies
+in parallel.
+
+See the Policy Server code [here](https://github.com/kubewarden/policy-server)
+
+#### Policies
+
+At the heart of Kubewarden's operational logic lies the Policies — OCI artifacts
```
Comment (jhkrug) [inline components.md:40]: Dunno, maybe core or center is better than heart?

File: components.md [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+[here](https://github.com/search?q=org%3Akubewarden%20topic%3Akubewarden-policy&type=repositories)
+
+#### ValidationWebhooks and MutatingWebhooks
+
+Kubewarden's integration with Kubernetes is orchestrated through
+ValidationWebhooks and MutatingWebhooks, mechanisms that funnel admission
+requests to Policy Servers. Created by the Kubewarden Controller, these
+webhooks are critical for directing traffic, ensuring that each admission
+request is evaluated against the relevant policy, facilitating a dynamic and
+automated policy enforcement process.
```
Comment (jhkrug) [inline components.md:56]:
```suggestion
Kubewarden's integration with Kubernetes is orchestrated through
ValidationWebhooks and MutatingWebhooks, mechanisms that funnel admission
requests to PolicyServers. Created by the Kubewarden Controller, these
webhooks are critical for directing traffic, ensuring that each admission
request is evaluated against the relevant policy, facilitating a dynamic and
automated policy enforcement process.
```

File: components.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+#### Audit Scanner
+
+The Audit Scanner introduces a proactive layer to Kubewarden’s security
+framework, periodically scanning the cluster to identify and report
+non-compliant resources. This component enhances cluster security by evaluating
+the cluster state against enforced policies, identifying discrepancies and
+facilitating timely remediation. It integrates seamlessly with Policy Servers,
+using existing policy definitions to ensure consistent enforcement and
+evaluation.
```
Comment (jhkrug) [inline components.md:66]: 'integrates' is sufficient, the 'seamlessly' is overkill

```suggestion
#### AuditScanner

The AuditScanner introduces a proactive layer to Kubewarden’s security
framework, periodically scanning the cluster to identify and report
non-compliant resources. This component enhances cluster security by evaluating
the cluster state against enforced policies, identifying discrepancies and
facilitating timely remediation. It integrates with Policy Servers,
using existing policy definitions to ensure consistent enforcement and
evaluation.
```

File: components.md [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+The Audit Scanner code can be found
```
Comment (jhkrug) [inline components.md:68]:
```suggestion
The AuditScanner code can be found
```

File: components.md [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+distinct function in the Kubewarden stack, ensuring users' compliance and
+secuirty policies are properly enforced.
+
+#### Kubewarden CRDs (Custom Resource Definitions)
+
+Kubewarden CRDs define resources for admission request validations. They detail
+the structure, not only for the `ClusterAdmissionPolicy` and `AdmissionPolicy`
+resources, which include policy configurations and matching rules for admission
+requests, but also for the `PolicyServer` resource. CRDs define the setup for
+Policy Servers, with deployment specific details and parameters.
```
Comment (jhkrug) [inline components.md:13]:
```suggestion
PolicyServers, with deployment specific details and parameters.
```

File: components.md [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+The Kubewarden CRDs are defined in the
+[kubewarden-controller](https://github.com/kubewarden/kubewarden-controller)
+repository.
+
+#### Kubewarden Controller
+
+The Kubewarden Controller is the conductor. It monitors Kubewarden CRDs and
+Kubernetes resources, ensuring the all resources necessary for running Policy
+Servers and policies are correctly configured.
```
Comment (jhkrug) [inline components.md:23]:
```suggestion
The Kubewarden Controller is the conductor. It monitors Kubewarden CRDs and
Kubernetes resources, ensuring the all resources necessary for running 
PolicyServers and policies are correctly configured. 
```

**Review by jhkrug — APPROVED** [review]
Great, a few minor suggestions.

I think we have decided on using PolicyServer as a preference in this thread: https://suse.slack.com/archives/C02DBSK7HC1/p1712234760652379 following the Kubernetes convention.

File: components.md (hunk 15 above)
Comment (jvanz) [inline components.md:40]:
```suggestion
At the core of Kubewarden's operational logic lies the Policies — OCI artifacts
```

**Pull request merged**